File: graph_shortest_path/a_start/src/AStartDemo.py
# Python
# グラフの最短経路: A-star
import heapq

class GraphData:
    def __init__(self):
        # 隣接ノードとその辺の重みを格納します。
        # キーは頂点、値はその頂点に隣接する頂点と重みの辞書です。
        # 例: { 'A': {'B': 1, 'C': 4}, 'B': {'A': 1, 'C': 2, 'D': 5}, ... }
        self._data = {}

    def get(self):
        # グラフの内部データを取得します。
        return self._data

    def get_vertices(self):
        # グラフの全頂点をリストとして返します。
        return list(self._data.keys())

    def get_edges(self):
        # グラフの全辺をリストとして返します。
        # 無向グラフの場合、(u, v, weight) の形式で返します。
        # 重複を避けるためにセットを使用します。
        edges = set()
        for vertex in self._data:
            for neighbor, weight in self._data[vertex]:
                # 辺を正規化してセットに追加 (小さい方の頂点を最初にするなど)
                edge = tuple(sorted((vertex, neighbor)))
                edges.add((edge[0], edge[1], weight)) # (u, v, weight) の形式で格納
        return list(edges)

    def get_neighbors(self, vertex):
        # 指定された頂点の隣接ノードと辺の重みのリストを返します。
        # 形式: [(隣接頂点, 重み), ...]
        if vertex in self._data:
            return self._data[vertex]
        else:
            return None # 頂点が存在しない場合はNoneを返す

    def add_vertex(self, vertex):
        # 新しい頂点をグラフに追加します。
        if vertex not in self._data:
            self._data[vertex] = []
            return True
        # 既に存在する場合は追加しないがTrueを返す（変更なしでも成功とみなす）
        return True

    def add_edge(self, vertex1, vertex2, weight):
        # 両頂点間に辺を追加します。重みを指定します。
        # 頂点がグラフに存在しない場合は追加します。
        if vertex1 not in self._data:
            self.add_vertex(vertex1)
        if vertex2 not in self._data:
            self.add_vertex(vertex2)
        
        # 両方向に辺を追加する（無向グラフ）
        # 既に同じ辺が存在する場合は重みを更新するかどうか？
        # 今回は単純に追加する（重複辺はプリム法では問題にならないが、データとしては綺麗でないかも）
        # ここでは、同じ頂点間の辺が存在しない場合のみ追加するように修正します。
        
        # vertex1 -> vertex2 の辺を追加（重み付き）
        edge_exists_v1v2 = False
        for i, (neighbor, _) in enumerate(self._data[vertex1]):
            if neighbor == vertex2:
                self._data[vertex1][i] = (vertex2, weight) # 既に存在する場合は重みを更新
                edge_exists_v1v2 = True
                break
        if not edge_exists_v1v2:
            self._data[vertex1].append((vertex2, weight))

        # vertex2 -> vertex1 の辺を追加（重み付き）
        edge_exists_v2v1 = False
        for i, (neighbor, _) in enumerate(self._data[vertex2]):
            if neighbor == vertex1:
                self._data[vertex2][i] = (vertex1, weight) # 既に存在する場合は重みを更新
                edge_exists_v2v1 = True
                break
        if not edge_exists_v2v1:
            self._data[vertex2].append((vertex1, weight))
        
        return True

    def clear(self):
        # グラフを空にします。
        self._data = {}
        return True

    def get_shortest_path(self, start_vertex, end_vertex, heuristic):
        if start_vertex not in self._data or end_vertex not in self._data:
            print("ERROR: 開始頂点または終了頂点がグラフに存在しません。")
            return None, float('inf')

        if start_vertex == end_vertex:
            return [start_vertex], 0

        # g_costs: 開始ノードから各ノードまでの既知の最短コスト
        g_costs = {vertex: float('inf') for vertex in self._data}
        g_costs[start_vertex] = 0

        # f_costs: g_costs + ヒューリスティックコスト (推定合計コスト)
        f_costs = {vertex: float('inf') for vertex in self._data}
        f_costs[start_vertex] = heuristic(start_vertex, end_vertex)

        # came_from: 最短経路で各ノードの直前のノードを記録
        came_from = {}

        # open_set: 探索すべきノードの優先度キュー (f_cost, node)
        open_set = [(f_costs[start_vertex], start_vertex)]

        # すでに open_set に追加されたノードを追跡するためのセット
        # (heapq は要素の存在チェックが遅いため) - 必要に応じて。
        # 今回はf_costをキーにした辞書を使う方が効率的かも。
        # あるいは、heapqに重複要素が入ることを許容し、取り出し時にg_cost/f_costが古いかチェックする。
        # 後者のアプローチで実装します。

        # open_set に (f_cost, g_cost, node) のタプルを入れることで、
        # f_cost が同じ場合の tie-breaking に g_cost を使えるが、
        # 基本的な A* は (f_cost, node) で十分。ここではシンプルに (f_cost, node) を使用。
        # heapq.heappush(open_set, (f_costs[start_vertex], g_costs[start_vertex], start_vertex)) # if using (f, g, node)

        while open_set:
            # open_set から最も f_cost が低いノードを取り出す
            current_f_cost, current_vertex = heapq.heappop(open_set)
            # current_f_cost, current_g_cost, current_vertex = heapq.heappop(open_set) # if using (f, g, node)

            # 取り出したノードの f_cost が、記録されている f_costs[current_vertex] より大きい場合、
            # それは古い情報なので無視して次のノードに進む
            # (heapq は要素を削除できないため、より低い f_cost で同じノードが後から追加される可能性がある)
            if current_f_cost > f_costs[current_vertex]:
                 continue

            # 目標ノードに到達した場合、経路を再構築して返す
            if current_vertex == end_vertex:
                return self._reconstruct_path(came_from, end_vertex), g_costs[end_vertex]

            # 現在のノードの隣接ノードを調べる
            neighbors = self.get_neighbors(current_vertex)
            if neighbors is None: # 孤立したノードの場合など
                continue

            for neighbor, weight in neighbors:
                # 現在のノードを経由した場合の隣接ノードへの新しい g_cost
                tentative_g_cost = g_costs[current_vertex] + weight

                # 新しい g_cost が、現在記録されている隣接ノードへの g_cost よりも小さい場合
                if tentative_g_cost < g_costs[neighbor]:
                    # 経路情報を更新
                    came_from[neighbor] = current_vertex
                    g_costs[neighbor] = tentative_g_cost
                    f_costs[neighbor] = g_costs[neighbor] + heuristic(neighbor, end_vertex)
                    
                    # 隣接ノードを open_set に追加（または優先度を更新）
                    # heapq は優先度更新を直接サポートしないため、新しいエントリを追加
                    # 古いエントリは取り出し時のチェックでスキップされる
                    heapq.heappush(open_set, (f_costs[neighbor], neighbor))
                    # heapq.heappush(open_set, (f_costs[neighbor], g_costs[neighbor], neighbor)) # if using (f, g, node)

        # open_set が空になっても目標ノードに到達しなかった場合、経路は存在しない
        return None, float('inf')

    def _reconstruct_path(self, came_from, current_vertex):
        path = []
        while current_vertex in came_from:
            path.append(current_vertex)
            current_vertex = came_from[current_vertex]
        path.append(current_vertex) # 開始ノードを追加
        path.reverse() # 経路を逆順にする (開始 -> 目標)
        return path

# ヒューリスティック関数 (この例では常に0、ダイクストラ法と同じ)
# 実際のA*では、問題に応じた適切な推定関数を使用する必要があります。
def dummy_heuristic(u, v):
    # u と v の間に何らかの推定距離を計算する関数
    # ここではダミーとして常に0を返す
    return 0

def main():
    print("A-start TEST -----> start")

    graph_data = GraphData()

    graph_data.clear()
    inputList = [('A', 'B', 4), ('B', 'C', 3), ('B', 'D', 2), ('D', 'A', 1), ('A', 'C', 2), ('B', 'D', 2)]
    for input in inputList:
        graph_data.add_edge(*input)
    print("\nグラフの頂点:", graph_data.get_vertices())
    print("グラフの辺 (重み付き):", graph_data.get_edges())
    input = ('A', 'B')
    shortest_path = graph_data.get_shortest_path(input[0], input[1], dummy_heuristic)
    print(f"経路{input[0]}-{input[1]} の最短経路は {shortest_path[0]} (重み: {shortest_path[1]})")

    graph_data.clear()
    inputList = [('A', 'B', 4), ('C', 'D', 4), ('E', 'F', 1), ('F', 'G', 1)]
    for input in inputList:
        graph_data.add_edge(*input)
    print("\nグラフの頂点:", graph_data.get_vertices())
    print("グラフの辺 (重み付き):", graph_data.get_edges())
    input = ('A', 'B')
    shortest_path = graph_data.get_shortest_path(input[0], input[1], dummy_heuristic)
    print(f"経路{input[0]}-{input[1]} の最短経路は {shortest_path[0]} (重み: {shortest_path[1]})")

    graph_data.clear()
    inputList = [('A', 'B', 4), ('B', 'C', 3), ('D', 'E', 5)]
    for input in inputList:
        graph_data.add_edge(*input)
    print("\nグラフの頂点:", graph_data.get_vertices())
    print("グラフの辺 (重み付き):", graph_data.get_edges())
    input = ('A', 'D')
    shortest_path = graph_data.get_shortest_path(input[0], input[1], dummy_heuristic)
    print(f"経路{input[0]}-{input[1]} の最短経路は {shortest_path[0]} (重み: {shortest_path[1]})")

    graph_data.clear()
    inputList = []
    for input in inputList:
        graph_data.add_edge(*input)
    print("\nグラフの頂点:", graph_data.get_vertices())
    print("グラフの辺 (重み付き):", graph_data.get_edges())
    input = ('A', 'B')
    shortest_path = graph_data.get_shortest_path(input[0], input[1], dummy_heuristic)
    print(f"経路{input[0]}-{input[1]} の最短経路は {shortest_path[0]} (重み: {shortest_path[1]})")

    print("\nA-start TEST <----- end")

if __name__ == "__main__":
    main()

File: graph_shortest_path/a_start/src/AStartDemo.go
// Go
// グラフの最短経路: A-star

package main

import (
	"container/heap"
	"fmt"
	"math"
)

// GraphData はグラフのデータ構造を表します
type GraphData struct {
	// 隣接ノードとその辺の重みを格納します
	// キーは頂点、値はその頂点に隣接する頂点と重みのスライスです
	data map[string][]Edge
}

// Edge は頂点と重みのペアを表します
type Edge struct {
	Neighbor string
	Weight   float64
}

// NewGraphData はGraphDataの新しいインスタンスを作成します
func NewGraphData() *GraphData {
	return &GraphData{
		data: make(map[string][]Edge),
	}
}

// Get はグラフの内部データを取得します
func (g *GraphData) Get() map[string][]Edge {
	return g.data
}

// GetVertices はグラフの全頂点をスライスとして返します
func (g *GraphData) GetVertices() []string {
	vertices := make([]string, 0, len(g.data))
	for vertex := range g.data {
		vertices = append(vertices, vertex)
	}
	return vertices
}

// GetEdges はグラフの全辺をスライスとして返します
// 無向グラフの場合、(u, v, weight) の形式で返します
func (g *GraphData) GetEdges() []struct {
	Vertex1, Vertex2 string
	Weight           float64
} {
	edges := make(map[string]struct {
		Vertex1, Vertex2 string
		Weight           float64
	})

	for vertex, neighbors := range g.data {
		for _, edge := range neighbors {
			// 辺を正規化して（小さい方の頂点を最初にするなど）
			v1, v2 := vertex, edge.Neighbor
			if v1 > v2 {
				v1, v2 = v2, v1
			}
			key := v1 + ":" + v2
			edges[key] = struct {
				Vertex1, Vertex2 string
				Weight           float64
			}{
				Vertex1: v1,
				Vertex2: v2,
				Weight:  edge.Weight,
			}
		}
	}

	result := make([]struct {
		Vertex1, Vertex2 string
		Weight           float64
	}, 0, len(edges))

	for _, edge := range edges {
		result = append(result, edge)
	}

	return result
}

// GetNeighbors は指定された頂点の隣接ノードと辺の重みのスライスを返します
func (g *GraphData) GetNeighbors(vertex string) []Edge {
	if neighbors, exists := g.data[vertex]; exists {
		return neighbors
	}
	return nil
}

// AddVertex は新しい頂点をグラフに追加します
func (g *GraphData) AddVertex(vertex string) bool {
	if _, exists := g.data[vertex]; !exists {
		g.data[vertex] = []Edge{}
	}
	return true
}

// AddEdge は両頂点間に辺を追加します。重みを指定します
func (g *GraphData) AddEdge(vertex1, vertex2 string, weight float64) bool {
	if _, exists := g.data[vertex1]; !exists {
		g.AddVertex(vertex1)
	}
	if _, exists := g.data[vertex2]; !exists {
		g.AddVertex(vertex2)
	}

	// vertex1 -> vertex2 の辺を追加（重み付き）
	edgeExistsV1V2 := false
	for i, edge := range g.data[vertex1] {
		if edge.Neighbor == vertex2 {
			g.data[vertex1][i] = Edge{Neighbor: vertex2, Weight: weight} // 既に存在する場合は重みを更新
			edgeExistsV1V2 = true
			break
		}
	}
	if !edgeExistsV1V2 {
		g.data[vertex1] = append(g.data[vertex1], Edge{Neighbor: vertex2, Weight: weight})
	}

	// vertex2 -> vertex1 の辺を追加（重み付き）
	edgeExistsV2V1 := false
	for i, edge := range g.data[vertex2] {
		if edge.Neighbor == vertex1 {
			g.data[vertex2][i] = Edge{Neighbor: vertex1, Weight: weight} // 既に存在する場合は重みを更新
			edgeExistsV2V1 = true
			break
		}
	}
	if !edgeExistsV2V1 {
		g.data[vertex2] = append(g.data[vertex2], Edge{Neighbor: vertex1, Weight: weight})
	}

	return true
}

// Clear はグラフを空にします
func (g *GraphData) Clear() bool {
	g.data = make(map[string][]Edge)
	return true
}

// HeuristicFunc は2つの頂点間の推定距離を計算する関数型
type HeuristicFunc func(string, string) float64

// PriorityQueue はA*アルゴリズムで使用する優先度キューを実装します
type PriorityQueue []*Item

// Item は優先度キューの要素です
type Item struct {
	Value    string  // 頂点の名前
	FCost    float64 // 推定の総コスト
	Priority int     // ヒープ内での優先度
	Index    int     // ヒープ内でのインデックス
}

// Len はPriorityQueueのヒープインターフェースを実装します
func (pq PriorityQueue) Len() int { return len(pq) }

// Less はPriorityQueueのヒープインターフェースを実装します
func (pq PriorityQueue) Less(i, j int) bool {
	// FCostが小さい方を優先します
	return pq[i].FCost < pq[j].FCost
}

// Swap はPriorityQueueのヒープインターフェースを実装します
func (pq PriorityQueue) Swap(i, j int) {
	pq[i], pq[j] = pq[j], pq[i]
	pq[i].Index = i
	pq[j].Index = j
}

// Push はPriorityQueueのヒープインターフェースを実装します
func (pq *PriorityQueue) Push(x interface{}) {
	n := len(*pq)
	item := x.(*Item)
	item.Index = n
	*pq = append(*pq, item)
}

// Pop はPriorityQueueのヒープインターフェースを実装します
func (pq *PriorityQueue) Pop() interface{} {
	old := *pq
	n := len(old)
	item := old[n-1]
	old[n-1] = nil  // メモリリークを避けるため
	item.Index = -1 // 安全のため
	*pq = old[0 : n-1]
	return item
}

// GetShortestPath はA*アルゴリズムを使用して最短経路を見つけます
func (g *GraphData) GetShortestPath(startVertex, endVertex string, heuristic HeuristicFunc) ([]string, float64) {
	if _, exists1 := g.data[startVertex]; !exists1 {
		fmt.Println("ERROR: 開始頂点がグラフに存在しません。")
		return nil, math.Inf(1)
	}
	if _, exists2 := g.data[endVertex]; !exists2 {
		fmt.Println("ERROR: 終了頂点がグラフに存在しません。")
		return nil, math.Inf(1)
	}

	if startVertex == endVertex {
		return []string{startVertex}, 0
	}

	// g_costs: 開始ノードから各ノードまでの既知の最短コスト
	gCosts := make(map[string]float64)
	for vertex := range g.data {
		gCosts[vertex] = math.Inf(1)
	}
	gCosts[startVertex] = 0

	// f_costs: g_costs + ヒューリスティックコスト (推定合計コスト)
	fCosts := make(map[string]float64)
	for vertex := range g.data {
		fCosts[vertex] = math.Inf(1)
	}
	fCosts[startVertex] = heuristic(startVertex, endVertex)

	// came_from: 最短経路で各ノードの直前のノードを記録
	cameFrom := make(map[string]string)

	// open_set: 探索すべきノードの優先度キュー
	openSet := &PriorityQueue{}
	heap.Init(openSet)
	heap.Push(openSet, &Item{
		Value: startVertex,
		FCost: fCosts[startVertex],
	})

	visited := make(map[string]bool) // 処理済みのノードを追跡

	for openSet.Len() > 0 {
		// open_set から最も f_cost が低いノードを取り出す
		current := heap.Pop(openSet).(*Item)
		currentVertex := current.Value

		// 取り出したノードの f_cost が現在の記録よりも大きければスキップ
		if current.FCost > fCosts[currentVertex] {
			continue
		}

		// 目標ノードに到達した場合、経路を再構築して返す
		if currentVertex == endVertex {
			return g.reconstructPath(cameFrom, endVertex), gCosts[endVertex]
		}

		// 処理済みとしてマーク
		visited[currentVertex] = true

		// 現在のノードの隣接ノードを調べる
		neighbors := g.GetNeighbors(currentVertex)
		if neighbors == nil {
			continue
		}

		for _, edge := range neighbors {
			neighbor := edge.Neighbor
			weight := edge.Weight

			// すでに処理済みのノードはスキップ
			if visited[neighbor] {
				continue
			}

			// 現在のノードを経由した場合の隣接ノードへの新しい g_cost
			tentativeGCost := gCosts[currentVertex] + weight

			// 新しい g_cost が、現在記録されている隣接ノードへの g_cost よりも小さい場合
			if tentativeGCost < gCosts[neighbor] {
				// 経路情報を更新
				cameFrom[neighbor] = currentVertex
				gCosts[neighbor] = tentativeGCost
				fCosts[neighbor] = gCosts[neighbor] + heuristic(neighbor, endVertex)

				// 隣接ノードを open_set に追加
				heap.Push(openSet, &Item{
					Value: neighbor,
					FCost: fCosts[neighbor],
				})
			}
		}
	}

	// open_set が空になっても目標ノードに到達しなかった場合、経路は存在しない
	return nil, math.Inf(1)
}

// reconstructPath は最短経路を再構築します
func (g *GraphData) reconstructPath(cameFrom map[string]string, currentVertex string) []string {
	path := []string{}
	for {
		path = append([]string{currentVertex}, path...) // 先頭に追加
		prev, exists := cameFrom[currentVertex]
		if !exists {
			break
		}
		currentVertex = prev
	}
	return path
}

// DummyHeuristic はダミーのヒューリスティック関数（常に0を返す、ダイクストラ法と同じ）
func DummyHeuristic(u, v string) float64 {
	// u と v の間に何らかの推定距離を計算する関数
	// ここではダミーとして常に0を返す
	return 0
}

func main() {
	fmt.Println("A-start TEST -----> start")

	graphData := NewGraphData()

	graphData.Clear()
	inputList := []struct {
		Vertex1, Vertex2 string
		Weight           float64
	}{
		{"A", "B", 4},
		{"B", "C", 3},
		{"B", "D", 2},
		{"D", "A", 1},
		{"A", "C", 2},
		{"B", "D", 2},
	}
	for _, input := range inputList {
		graphData.AddEdge(input.Vertex1, input.Vertex2, input.Weight)
	}
	fmt.Println("\nグラフの頂点:", graphData.GetVertices())
	fmt.Println("グラフの辺 (重み付き):", graphData.GetEdges())
	input := [2]string{"A", "B"}
	shortestPath, weight := graphData.GetShortestPath(input[0], input[1], DummyHeuristic)
	fmt.Printf("経路%s-%s の最短経路は %v (重み: %v)\n", input[0], input[1], shortestPath, weight)

	graphData.Clear()
	inputList = []struct {
		Vertex1, Vertex2 string
		Weight           float64
	}{
		{"A", "B", 4},
		{"C", "D", 4},
		{"E", "F", 1},
		{"F", "G", 1},
	}
	for _, input := range inputList {
		graphData.AddEdge(input.Vertex1, input.Vertex2, input.Weight)
	}
	fmt.Println("\nグラフの頂点:", graphData.GetVertices())
	fmt.Println("グラフの辺 (重み付き):", graphData.GetEdges())
	input = [2]string{"A", "B"}
	shortestPath, weight = graphData.GetShortestPath(input[0], input[1], DummyHeuristic)
	fmt.Printf("経路%s-%s の最短経路は %v (重み: %v)\n", input[0], input[1], shortestPath, weight)

	graphData.Clear()
	inputList = []struct {
		Vertex1, Vertex2 string
		Weight           float64
	}{
		{"A", "B", 4},
		{"B", "C", 3},
		{"D", "E", 5},
	}
	for _, input := range inputList {
		graphData.AddEdge(input.Vertex1, input.Vertex2, input.Weight)
	}
	fmt.Println("\nグラフの頂点:", graphData.GetVertices())
	fmt.Println("グラフの辺 (重み付き):", graphData.GetEdges())
	input = [2]string{"A", "D"}
	shortestPath, weight = graphData.GetShortestPath(input[0], input[1], DummyHeuristic)
	fmt.Printf("経路%s-%s の最短経路は %v (重み: %v)\n", input[0], input[1], shortestPath, weight)

	graphData.Clear()
	inputList = []struct {
		Vertex1, Vertex2 string
		Weight           float64
	}{}
	for _, input := range inputList {
		graphData.AddEdge(input.Vertex1, input.Vertex2, input.Weight)
	}
	fmt.Println("\nグラフの頂点:", graphData.GetVertices())
	fmt.Println("グラフの辺 (重み付き):", graphData.GetEdges())
	input = [2]string{"A", "B"}
	shortestPath, weight = graphData.GetShortestPath(input[0], input[1], DummyHeuristic)
	fmt.Printf("経路%s-%s の最短経路は %v (重み: %v)\n", input[0], input[1], shortestPath, weight)

	fmt.Println("\nA-start TEST <----- end")
}
